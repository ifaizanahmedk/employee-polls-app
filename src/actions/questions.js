import { saveQuestion, saveQuestionAnswer } from "../utils/api";
import { addAnswerUser, addQuestionUser } from "./users";

export const ADD_QUESTION = "ADD_QUESTION";
export const RECEIVE_QUESTIONS = "RECEIVE_QUESTIONS";
export const SAVE_VOTE = "SAVE_VOTE";

export function addQuestion(question) {
	return {
		type: ADD_QUESTION,
		question,
	};
}

export function receiveQuestions(questions) {
	return {
		type: RECEIVE_QUESTIONS,
		questions,
	};
}

export function saveVote(authedUser, qid, answer) {
	return {
		type: SAVE_VOTE,
		authedUser,
		qid,
		answer,
	};
}

export function handleAddQuestion(firstOption, secondOption) {
	return (dispatch, getState) => {
		const { authedUser } = getState();
		return saveQuestion({
			optionOneText: firstOption,
			optionTwoText: secondOption,
			author: authedUser,
		}).then((question) => {
			dispatch(addQuestion(question));
			dispatch(addQuestionUser(question));
		});
	};
}

export function handleVote(qid, answer) {
	return (dispatch, getState) => {
		const { authedUser } = getState();

		return saveQuestionAnswer({ authedUser, qid, answer }).then(() => {
			dispatch(saveVote(authedUser, qid, answer));
			dispatch(addAnswerUser(authedUser, qid, answer));
		});
	};
}
