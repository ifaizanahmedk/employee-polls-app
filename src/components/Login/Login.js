import { useState } from "react";
import { connect } from "react-redux";
import { useNavigate } from "react-router-dom";
import setAuthedUser from "../../actions/authedUser";
import votingBanner from "../../images/lets-vote.png";
import { Button, Col, FloatingLabel, Form, Image, Row } from "react-bootstrap";

const Login = ({ dispatch, users }) => {
	const navigate = useNavigate();
	const [selectedUser, setSelectedUser] = useState({});
	const [password, setPassword] = useState("");
	const [error, setError] = useState(false);

	const handleChange = (e) => {
		switch (e.target.id) {
			case "user-select":
				setSelectedUser(e.target.value);
				return;

			case "user-password":
				setPassword(e.target.value);
				return;

			default:
				return;
		}
	};

	const handleSubmit = (e) => {
		e.preventDefault();

		let userPassword = users[selectedUser].password;

		if (userPassword === password) {
			dispatch(setAuthedUser(selectedUser));
		} else {
			setError(true);
		}
	};

	return (
		<>
			<Row className="m-5">
				<Col md="12">
					<h1 className="text-center">Employee Polls App</h1>
				</Col>
			</Row>
			<Row>
				<Col md="6">
					<Image src={votingBanner} width={350} fluid={true} />
				</Col>
				<Col md="6">
					<Form name="user-login-form" className="login-form ml-3">
						<Form.Group className="mb-3">
							<h3>Login</h3>
						</Form.Group>
						<Form.Group className="mb-3">
							<FloatingLabel
								controlId="user-select"
								label="Select your User">
								<Form.Select
									data-testid="user-select"
									aria-label="Select your User"
									size="lg"
									defaultValue="default"
									onChange={handleChange}>
									<option value="default" disabled>
										What's your user?
									</option>
									{users &&
										Object.values(users).map((user) => (
											<option
												data-testid={user.id}
												key={user.id}
												value={user.id}>
												{user.id}
											</option>
										))}
								</Form.Select>
							</FloatingLabel>
						</Form.Group>
						<Form.Group className="mb-3">
							<FloatingLabel
								controlId="user-password"
								label="Password">
								<Form.Control
									data-testid="user-password"
									size="lg"
									type="password"
									placeholder="Password"
									onChange={handleChange}
									value={password}
									isInvalid={error}
								/>
							</FloatingLabel>
							{error && (
								<small
									data-testid="error-message"
									className="text-danger">
									Incorrect Password. Try again!
								</small>
							)}
						</Form.Group>

						<Form.Group className="mb-3">
							<Button
								variant="primary"
								type="submit"
								className="login-form-button"
								id="login-form-button"
								data-testid="login-form-button"
								disabled={!selectedUser || !password}
								onClick={handleSubmit}>
								Log in
							</Button>
						</Form.Group>
					</Form>
				</Col>
			</Row>
		</>
	);
};

const mapStateToProps = ({ users }) => ({ users });

export default connect(mapStateToProps)(Login);
