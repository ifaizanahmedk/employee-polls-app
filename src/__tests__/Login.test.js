// import "@testing-library/jest-dom";
import { store } from "../store";
import { Provider } from "react-redux";
import { BrowserRouter as Router } from "react-router-dom";
import { fireEvent, render, waitFor } from "@testing-library/react";
import Login from "../components/Login/Login";
import { handleInitialData } from "../actions/shared";
import userEvent from "@testing-library/user-event";

describe("Login", () => {
	it("should render the component", async () => {
		const component = render(
			<Provider store={store}>
				<Router>
					<Login />
				</Router>
			</Provider>
		);
		expect(component).toBeDefined();
		expect(component).toMatchSnapshot();
	});

	it("will verify that the login component has a login field, a password field and a submit button", () => {
		const component = render(
			<Provider store={store}>
				<Router>
					<Login />
				</Router>
			</Provider>
		);

		const userSelect = component.getByTestId("user-select");
		const passwordInput = component.getByTestId("user-password");
		const loginButton = component.getByTestId("login-form-button");

		expect(userSelect).toBeInTheDocument();
		expect(passwordInput).toBeInTheDocument();
		expect(loginButton).toBeInTheDocument();
	});

	it("will return Error if the username and password doesn't match", async () => {
		await store.dispatch(handleInitialData());

		const component = render(
			<Provider store={store}>
				<Router>
					<Login />
				</Router>
			</Provider>
		);

		const userSelect = component.getByTestId("user-select");
		const password = component.getByTestId("user-password");
		const loginButton = component.getByTestId("login-form-button");

		const userId = await waitFor(() => component.getByTestId("sarahedo"), {
			timeout: 2000,
		}); //we wait extra in order to have the users after the dispatch (1000ms)

		userEvent.selectOptions(
			component.getByTestId("user-select"),
			"sarahedo"
		);
		expect(component.getByTestId("sarahedo").selected).toBe(true);

		fireEvent.change(userSelect);
		fireEvent.click(userId);
		fireEvent.change(password, { target: { value: "wrong_password" } });
		fireEvent.click(loginButton);

		expect(loginButton).toBeInTheDocument();
		// makes sure error message is in the DOM after error happens.
		expect(component.getByTestId("error-message")).toBeInTheDocument();
	});
});
