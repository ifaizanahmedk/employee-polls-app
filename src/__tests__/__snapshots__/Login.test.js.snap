// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should render the component 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="m-5 row"
      >
        <div
          class="col-md-12"
        >
          <h1
            class="text-center"
          >
            Employee Polls App
          </h1>
        </div>
      </div>
      <div
        class="row"
      >
        <div
          class="col-md-6"
        >
          <img
            class="img-fluid"
            src="lets-vote.png"
            width="350"
          />
        </div>
        <div
          class="col-md-6"
        >
          <form
            class="login-form ml-3"
            name="user-login-form"
          >
            <div
              class="mb-3"
            >
              <h3>
                Login
              </h3>
            </div>
            <div
              class="mb-3"
            >
              <div
                class="form-floating"
              >
                <select
                  aria-label="Select your User"
                  class="form-select form-select-lg"
                  data-testid="user-select"
                  id="user-select"
                >
                  <option
                    disabled=""
                    selected=""
                    value="default"
                  >
                    What's your user?
                  </option>
                </select>
                <label
                  for="user-select"
                >
                  Select your User
                </label>
              </div>
            </div>
            <div
              class="mb-3"
            >
              <div
                class="form-floating"
              >
                <input
                  class="form-control form-control-lg"
                  data-testid="user-password"
                  id="user-password"
                  placeholder="Password"
                  type="password"
                  value=""
                />
                <label
                  for="user-password"
                >
                  Password
                </label>
              </div>
            </div>
            <div
              class="mb-3"
            >
              <button
                class="login-form-button btn btn-primary"
                data-testid="login-form-button"
                disabled=""
                id="login-form-button"
                type="submit"
              >
                Log in
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="m-5 row"
    >
      <div
        class="col-md-12"
      >
        <h1
          class="text-center"
        >
          Employee Polls App
        </h1>
      </div>
    </div>
    <div
      class="row"
    >
      <div
        class="col-md-6"
      >
        <img
          class="img-fluid"
          src="lets-vote.png"
          width="350"
        />
      </div>
      <div
        class="col-md-6"
      >
        <form
          class="login-form ml-3"
          name="user-login-form"
        >
          <div
            class="mb-3"
          >
            <h3>
              Login
            </h3>
          </div>
          <div
            class="mb-3"
          >
            <div
              class="form-floating"
            >
              <select
                aria-label="Select your User"
                class="form-select form-select-lg"
                data-testid="user-select"
                id="user-select"
              >
                <option
                  disabled=""
                  selected=""
                  value="default"
                >
                  What's your user?
                </option>
              </select>
              <label
                for="user-select"
              >
                Select your User
              </label>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <div
              class="form-floating"
            >
              <input
                class="form-control form-control-lg"
                data-testid="user-password"
                id="user-password"
                placeholder="Password"
                type="password"
                value=""
              />
              <label
                for="user-password"
              >
                Password
              </label>
            </div>
          </div>
          <div
            class="mb-3"
          >
            <button
              class="login-form-button btn btn-primary"
              data-testid="login-form-button"
              disabled=""
              id="login-form-button"
              type="submit"
            >
              Log in
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
